### DISCUSSION TITLE
Hyundai Ioniq Electric 2017

### DISCUSSION THREAD

**Discussion by drcvach** [meatpiHQ/wican-fw#75] [Q&A] [answered]
Hello,
I'm trying to get reading values from the first generation Ioniq to work.
I have successfully managed to find a way to parse messages as described in "Spreadsheet_IoniqEV_BMS_2101_2105.xls" from https://github.com/JejuSoul/OBD-PIDs-for-HKMC-EVs/tree/master/Ioniq%20EV%20-%2028kWh.
Unfortunately, I ran into two problems that I haven't been able to overcome yet.

**1)** 
With Ioniq, the data is stored in two PIDs: 2101 and 2105. According to the mentioned xls (which corresponds to reality in the distribution of parameters in the data part of the RX packets), only the first byte, i.e. B0, is static out of the 8 bytes of the payload (B0 byte represents the line number). The other bytes are used to encode Ioniq's parameters.
But according to the experiments I made, "CAN to JSON interpreter" in WiCAN Settings page needs B2 as PID. Only in this case it produces the rx data to MQTT.
Here is an example for **2105** (SoH extraction):
{"bus":"0","type":"tx","ts":39627,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[2,33,**5**,170,170,170,170,170]}]}
	{"bus":"0","type":"rx","ts":39637,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[16,45,97,**5**,255,255,255,255]}]}
{"bus":"0","type":"tx","ts":39639,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[48,0,10,170,170,170,170,170]}]}
	{"bus":"0","type":"rx","ts":39652,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[33,0,0,0,0,0,14,14]}]}
	{"bus":"0","type":"rx","ts":39667,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[34,15,13,13,14,14,37,225]}]}
	{"bus":"0","type":"rx","ts":39686,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[35,38,72,0,1,100,12,13]}]}
	{"bus":"0","type":"rx","ts":39697,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[36,3,232,66,3,232,79,164]}]}
	{"bus":"0","type":"rx","ts":39712,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[37,0,49,0,0,0,0,0]}]}
	{"bus":"0","type":"rx","ts":39727,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[38,0,0,0,0,0,0,0]}]}
For example, on the penultimate line, i.e. 36, the SoH battery is stored in B1 and B2. But I have to save B2=232 as PID to "CAN to JSON interpreter" for the translation to work.
The PID is stored only in the first rx packet at position B3 as the value 5: [16,45,97,**5**,255,255,255,255].
Another complication is that PID 2101 also uses the same structure, so the fact that it is 2101 versus 2105 can only be known from the first two packets.
Example **2101**:
{"bus":"0","type":"tx","ts":10928,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[2,33,**1**,170,170,170,170,170]}]}
	{"bus":"0","type":"rx","ts":10936,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[16,61,97,**1**,255,255,255,255]}]}
{"bus":"0","type":"tx","ts":10938,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[48,0,10,170,170,170,170,170]}]}
	{"bus":"0","type":"rx","ts":10952,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[33,139,38,72,38,72,163,255]}]}
	{"bus":"0","type":"rx","ts":10966,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[34,232,14,87,13,11,13,13]}]}
	{"bus":"0","type":"rx","ts":10986,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[35,12,11,13,0,11,191,34]}]}
	{"bus":"0","type":"rx","ts":11002,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[36,191,1,0,0,144,0,11]}]}
	{"bus":"0","type":"rx","ts":11016,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[37,48,198,0,11,50,41,0]}]}
	{"bus":"0","type":"rx","ts":11032,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[38,4,23,202,0,3,246,251]}]}
	{"bus":"0","type":"rx","ts":11046,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[39,2,33,59,211,9,1,110]}]}
	{"bus":"0","type":"rx","ts":11062,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[40,0,0,0,0,3,232,0]}]}
Is there a way to deal with this?

**2)**
So far, I have not been able to find such a combination of packets that I would use to request data from the Ioniq via WiCAN. I send packets from Home Assistant or MQTT Explorer just fine (they appear in MQTT):
{"bus":"0","type":"tx","ts":39627,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd" :false,"data":[2,33,5,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":39639,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd" :false,"data":[48,0,10,170,170,170,170,170]}]}
to "wican/1091a83e9dd5/can/tx" topic, but no rx packets are returned. These are saved only when the "Car Scanner" communicates with WiCAN.
I tried to send packets to start the communication in advance, but so far without success.
Car Scanner communication examples are in the attachment.

[001_connect_succcessfull.txt](https://github.com/meatpiHQ/wican-fw/files/13302328/001_connect_succcessfull.txt)
[002_stay_connected_no_data_pulled.txt](https://github.com/meatpiHQ/wican-fw/files/13302329/002_stay_connected_no_data_pulled.txt)
[003_pulling_SoC_78and78,5_percent.txt](https://github.com/meatpiHQ/wican-fw/files/13302330/003_pulling_SoC_78and78.5_percent.txt)

Any ideas where is the problem?
Thank you very much in advance,
Petr

**Comment by meatpiHQ** [comment]
@drcvach Looks like this PID request has a multi frame response, currently the "CAN to JSON interpreter" cannot handle this type of requests. I think I can implement it but it will take some time. 


> to "wican/1091a83e9dd5/can/tx" topic, but no rx packets are returned. These are saved only when the "Car Scanner" communicates with WiCAN.

Is the "MQTT elm327 log" enabled? if it is then please disable it, you should see at least one response on "wican/1091a83e9dd5/can/tx" the response will be:
{"bus":"0","type":"rx","ts":39637,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[16,45,97,5,255,255,255,255]}]}

Then if you send:
{"bus":"0","type":"tx","ts":10938,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[48,0,10,170,170,170,170,170]}]}

you will receive the rest of the frames.

**Comment by drcvach** [comment]
@meatpiHQ  - Thank you for the tip. 
when I post these two TX packets from HA to MQTT:
   {"bus":"0","type":"tx","ts":29535,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[2,33,5,170,170,170,170,170]}]}
   {"bus":"0","type":"tx","ts":29547,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[48,0,10,170,170,170,170,170]}]}
and after adding these lines to "CAN to JSON interpreter" :
2028	SoC_display	232	0	1	B7/2		1000
2028	SoH	232	0	1	((B1*256)+B2)/10	1000
2028	BattHeater1Temp	72	0	1	B6		1000
2028	BattHeater2Temp	72	0	1	B7		1000
I am receiving decoded values via MQTT. That is great Thank you. Luckily SoC's PID (which I need the most) on B2 position is part of SoH which should not change fast, so I can live with that for some time.

But what still causes me problems is starting CAN communication.
If I record communication with the Ioniq through the "MQTT elm327 log", the following packets flow through the CarScanner app for login and subsequent SoC request:
{"bus":"0","type":"tx","ts":23866,"frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,0,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":24560,"frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,0,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":25399,"frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,32,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":26104,"frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,64,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":27354,"frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,0,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":28094,"frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,9,2,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":28720,"frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,0,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":29535,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[2,33,5,170,170,170,170,170]}]}
{"bus":"0","type":"tx","ts":29547,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[48,0,10,170,170,170,170,170]}]}

But when I send them in the same order vi HA, I do not get the RX response.
To be sure, I also tried 5ms pauses between packets, but without success. During this communication, is it necessary to actively parse rx packets and send tx only after receiving the correct response? Or is the error in something else?
P.

**Comment by meatpiHQ** [comment]
@drcvach Yes, you need to wait for the response before sending the next request.

**Comment by drcvach** [comment]
@meatpiHQ 
Thanks for the reply. Then, what is the best strategy for two-way communication with the car using HomeAssistant and "CAN to JSON interpreter"?
1) currently only the interpreted SoC value comes to me on MQTT thanks to the "CAN to JSON interpreter"
2) MQTT retrieval speed from HomeAssistant is slow,  I guess some messages can timeout before HA picks them up.
For the reasons mentioned above, the MQTT route does not seem appropriate to me, so is it necessary to use another method of communication?

**Comment by chris58008** [comment]
I have the same problems with the Hyundai Kona 2019 looks like the same data format.
Any news on a update for the JSON interpreter?

example: Cell voltage 1-32 the voltage is all the "193" = 193/50 = 3,86v 
`{"bus":"0","type":"tx","ts":43850,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[3,34,1,2,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":43864,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[16,39,98,1,2,255,255,255]}]}
{"bus":"0","type":"tx","ts":43873,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[48,0,10,170,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":43877,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[33,255,193,193,193,193,193,193]}]}
{"bus":"0","type":"rx","ts":43892,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[34,193,193,193,193,193,193,193]}]}
{"bus":"0","type":"rx","ts":43910,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[35,193,193,193,193,193,193,193]}]}
{"bus":"0","type":"rx","ts":43922,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[36,193,193,193,193,193,193,193]}]}
{"bus":"0","type":"rx","ts":43938,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[37,193,193,193,193,193,170,170]}]}`

**Comment by meatpiHQ** [comment]
@chris58008 There's no new update yet to address this type of messages. What needs to happen here, is to create an auto response for flow control. Hopeful I will be able to put some time into this soon.

**Reply by chris58008** [reply]
Maybe a quick fix would be to add a field to sort based on the BYTE0 similar to the PID field BYTE2 ?

That would be enough for me use the device :)

**Reply by drcvach** [reply]
Yes, most of us need SoC to control charging therefore BYTE0 filter would work when calling only 2105 to get SoC.
It worked for me for SoC extraction on Ioniq 28kWh, but Kona/eNiro have different byte structure and SoC is overwritten by other packets :-(

**Reply by meatpiHQ** [reply]
Actually a more generic approach would be to add a field to specify the byte index (1 to 8)

@drcvach I haven't forgotten this, but I'm knee deep in a project racing the Chinese new year. Hopefully I'll get it done soon.

**Reply by meatpiHQ** [reply, marked as answer]
@drcvach Please try this new version, I've added the index for PID. For the normal PID you should set the index to 2.

Here's how to use it:

Lets say you want to evaluate this frame, you set the index to "0" and PID to "38", everything else is as usual.
{"bus":"0","type":"rx","ts":11032,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[38,4,23,202,0,3,246,251]}]}


[wican-fw_v292_hv300.zip](https://github.com/meatpiHQ/wican-fw/files/14126390/wican-fw_v292_hv300.zip)

**Reply by chris58008** [reply]
Thanks alot 
![image](https://github.com/meatpiHQ/wican-fw/assets/85893534/c0d31caa-c52e-4cf0-8775-a49b0287f1e5)
This works !

**Reply by drcvach** [reply]
Yes, thank you very much - it works!

**Reply by brodykenrick** [reply]
FYI: I followed the above and got this working on an Ioniq5. There are a few bytes in the 2101 response that have shifted a little from Ioniq to Ioniq5.

Details over in the Home Assistant forum.

https://community.home-assistant.io/t/wican-esp32-based-open-source-obd2-obd-ii-car-diagnostics-can-bus-to-wifi-ble-usb-adapter-home-assistant-integration/456098/33?u=brodykenrick

**Comment by chrisgilldc** [comment]
Was pointed to this thread from the Discord. I have a 2022 Kona EV and am interested in getting CAN-JSON (and MQTT) working on it. Would be happy to help test.

**Reply by meatpiHQ** [reply]
@chrisgilldc Have a look at this link, you might find it useful

https://github.com/meatpiHQ/wican-fw/wiki/EV-Battery-SoC-in-Home-Assistant-%E2%80%90-Example

**Comment by Algram** [comment]
@drcvach I also have a Hyundai Ioniq 2017. Could you post a screeshot of your config? I am struggling to read the SOC.

The Car Scanner App is working, but only if the ignition is on. Is this normal? I need it to work with the car locked.

Thanks!

**Reply by drcvach** [reply]
Hi Algram, I sold the Ioniq but from my previous comments it should be something like this:
on older wican firmware like this:
   2028 SoC_display 232 0 1 B7/2 1000
and on newer one
   2028 SoC_display 36 0 48 8 V/2 1000

I think that Ignition has to be on or Ioniq's battery has to be charged.
Regards,
Petr

**Comment by djungelola** [comment]
> Hello, I'm trying to get reading values from the first generation Ioniq to work. I have successfully managed to find a way to parse messages as described in "Spreadsheet_IoniqEV_BMS_2101_2105.xls" from https://github.com/JejuSoul/OBD-PIDs-for-HKMC-EVs/tree/master/Ioniq%20EV%20-%2028kWh. Unfortunately, I ran into two problems that I haven't been able to overcome yet.
> 
> **1)** With Ioniq, the data is stored in two PIDs: 2101 and 2105. According to the mentioned xls (which corresponds to reality in the distribution of parameters in the data part of the RX packets), only the first byte, i.e. B0, is static out of the 8 bytes of the payload (B0 byte represents the line number). The other bytes are used to encode Ioniq's parameters. But according to the experiments I made, "CAN to JSON interpreter" in WiCAN Settings page needs B2 as PID. Only in this case it produces the rx data to MQTT. Here is an example for **2105** (SoH extraction): {"bus":"0","type":"tx","ts":39627,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[2,33,**5**,170,170,170,170,170]}]} {"bus":"0","type":"rx","ts":39637,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[16,45,97,**5**,255,255,255,255]}]} {"bus":"0","type":"tx","ts":39639,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[48,0,10,170,170,170,170,170]}]} {"bus":"0","type":"rx","ts":39652,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[33,0,0,0,0,0,14,14]}]} {"bus":"0","type":"rx","ts":39667,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[34,15,13,13,14,14,37,225]}]} {"bus":"0","type":"rx","ts":39686,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[35,38,72,0,1,100,12,13]}]} {"bus":"0","type":"rx","ts":39697,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[36,3,232,66,3,232,79,164]}]} {"bus":"0","type":"rx","ts":39712,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[37,0,49,0,0,0,0,0]}]} {"bus":"0","type":"rx","ts":39727,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[38,0,0,0,0,0,0,0]}]} For example, on the penultimate line, i.e. 36, the SoH battery is stored in B1 and B2. But I have to save B2=232 as PID to "CAN to JSON interpreter" for the translation to work. The PID is stored only in the first rx packet at position B3 as the value 5: [16,45,97,**5**,255,255,255,255]. Another complication is that PID 2101 also uses the same structure, so the fact that it is 2101 versus 2105 can only be known from the first two packets. Example **2101**: {"bus":"0","type":"tx","ts":10928,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[2,33,**1**,170,170,170,170,170]}]} {"bus":"0","type":"rx","ts":10936,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[16,61,97,**1**,255,255,255,255]}]} {"bus":"0","type":"tx","ts":10938,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd":false,"data":[48,0,10,170,170,170,170,170]}]} {"bus":"0","type":"rx","ts":10952,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[33,139,38,72,38,72,163,255]}]} {"bus":"0","type":"rx","ts":10966,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[34,232,14,87,13,11,13,13]}]} {"bus":"0","type":"rx","ts":10986,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[35,12,11,13,0,11,191,34]}]} {"bus":"0","type":"rx","ts":11002,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[36,191,1,0,0,144,0,11]}]} {"bus":"0","type":"rx","ts":11016,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[37,48,198,0,11,50,41,0]}]} {"bus":"0","type":"rx","ts":11032,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[38,4,23,202,0,3,246,251]}]} {"bus":"0","type":"rx","ts":11046,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[39,2,33,59,211,9,1,110]}]} {"bus":"0","type":"rx","ts":11062,"frame":[{"id":2028,"dlc":8,"rtr":false,"extd":false,"data":[40,0,0,0,0,3,232,0]}]} Is there a way to deal with this?
> 
> **2)** So far, I have not been able to find such a combination of packets that I would use to request data from the Ioniq via WiCAN. I send packets from Home Assistant or MQTT Explorer just fine (they appear in MQTT): {"bus":"0","type":"tx","ts":39627,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd" :false,"data":[2,33,5,170,170,170,170,170]}]} {"bus":"0","type":"tx","ts":39639,"frame":[{"id":2020,"dlc":8,"rtr":false,"extd" :false,"data":[48,0,10,170,170,170,170,170]}]} to "wican/1091a83e9dd5/can/tx" topic, but no rx packets are returned. These are saved only when the "Car Scanner" communicates with WiCAN. I tried to send packets to start the communication in advance, but so far without success. Car Scanner communication examples are in the attachment.
> 
> [001_connect_succcessfull.txt](https://github.com/meatpiHQ/wican-fw/files/13302328/001_connect_succcessfull.txt) [002_stay_connected_no_data_pulled.txt](https://github.com/meatpiHQ/wican-fw/files/13302329/002_stay_connected_no_data_pulled.txt) [003_pulling_SoC_78and78,5_percent.txt](https://github.com/meatpiHQ/wican-fw/files/13302330/003_pulling_SoC_78and78.5_percent.txt)
> 
> Any ideas where is the problem? Thank you very much in advance, Petr

Hi, was 1) in this question solved? Sitting with the same problem when wanting to get info from diffrent PIDS from my Kona but where the responses looks identical.

Br Ola

**Reply by chris58008** [reply]
https://github.com/meatpiHQ/wican-fw/discussions/86

**Comment by djungelola** [comment]
Yes, that part is under control. But the issue is that when sending different PID's in TX i get the same response structure in RX.
As example: 
2028	SOH		36	0	0	1	((B2*256)+B3)/10	1000
Depending on the TX PID the B2 and B3 will be different, so the sensor value will change for each TX and only one is the correct.
This means that i can only retrieve data from one PID.
